src/link/Wasm.zig

```diff
@@ -81,57 +138,112 @@ pub fn createEmpty(gpa: *Allocator, options: link.Options) !*Wasm {
+    }
+    for (self.data.segments.items) |entry| {
+        // decl's that live in data section do not generate idx_refs or func types
+        entry.decl.fn_link.wasm.code.deinit(self.base.allocator);
     }
     self.funcs.deinit(self.base.allocator);
     self.ext_funcs.deinit(self.base.allocator);
+    self.data.segments.deinit(self.base.allocator);
 }
 
-// Generate code for the Decl, storing it in memory to be later written to
-// the file on flush().
-pub fn updateDecl(self: *Wasm, module: *Module, decl: *Module.Decl) !void {
+pub fn allocateDeclIndexes(self: *Wasm, decl: *Module.Decl) !void {
     const typed_value = decl.typed_value.most_recent.typed_value;
-    if (typed_value.ty.zigTypeTag() != .Fn)
-        return error.TODOImplementNonFnDeclsForWasm;
 
-    if (decl.fn_link.wasm) |*fn_data| {
-        fn_data.functype.items.len = 0;
-        fn_data.code.items.len = 0;
-        fn_data.idx_refs.items.len = 0;
-    } else {
-        decl.fn_link.wasm = .{};
-        // dependent on function type, appends it to the correct list
-        switch (decl.typed_value.most_recent.typed_value.val.tag()) {
+    switch (typed_value.ty.zigTypeTag()) {
+        .Array => {
+            // if the codegen of the given decl contributes to the data segment
+            // we must calculate its data length now so that the data offsets are available
+            // to other decls when called
+            const data_len = self.calcDataLen(typed_value);
+            try self.data.segments.append(self.base.allocator, .{
+                .decl = decl,
+                .data = undefined,
+                .len = data_len,
+            });
+
+            // detect if we can replace it into a to-be-deleted decl's spot to ensure no gaps are
+            // made in our data segment
+            const idx: ?usize = for (self.data.segments.items) |entry, i| {
+                if (entry.decl.deletion_flag) break i;
+            } else null;
+
+            if (idx) |id| {
+                // swap to-be-removed decl with newly added to create a contigious valid data segment
+                const items = self.data.segments.items;
+                std.mem.swap(
+                    std.meta.Child(@TypeOf(items)),
+                    &items[id],
+                    &items[items.len - 1],
+                );
+            }
+        },
+        .Fn => if (self.getFuncidx(decl) == null) switch (typed_value.val.tag()) {
+            // dependent on function type, appends it to the correct list
             .function => try self.funcs.append(self.base.allocator, decl),
             .extern_fn => try self.ext_funcs.append(self.base.allocator, decl),
-            else => return error.TODOImplementNonFnDeclsForWasm,
-        }
+            else => unreachable,
+        },
+        else => {},
+    }
+}
+
+/// Calculates the length of the data segment that will be occupied by the given `TypedValue`
+fn calcDataLen(self: *Wasm, typed_value: TypedValue) u32 {
```

Comment:
Let's chat about this too - I don't quite understand why this mechanism is needed.